src/VisualStudio/Core/Def/Implementation/EditAndContinue/VisualStudioActiveStatementProvider.cs

```diff
@@ -0,0 +1,118 @@
+                                    {
+                                        if (cancellationToken.IsCancellationRequested)
+                                        {
+                                            workList.Cancel();
+                                            completion.SetCanceled();
+                                        }
+
+                                        if (sourcePositionResult.ErrorCode == 0)
+                                        {
+                                            builders[runtimeIndex][index] = new ActiveStatementDebugInfo(
+                                                instructionId,
+                                                sourcePositionResult.SourcePosition.DocumentName,
+                                                ToLinePositionSpan(sourcePositionResult.SourcePosition.TextSpan),
+                                                threads.ToImmutableAndFree(),
+                                                flags);
+                                        }
+
+                                        // the last active statement of the current runtime has been processed:
+                                        if (Interlocked.Decrement(ref pendingStatements) == 0)
+                                        {
+                                            instructionMap.Free();
+
+                                            // the last active statement of the last runtime has been processed:
+                                            if (Interlocked.Decrement(ref pendingRuntimes) == 0)
+                                            {
+                                                completion.SetResult(builders.ToImmutableArrayAndFree());
+                                            }
+                                        }
+                                    });
+                                }
+                            });
+
+                            runtimeCount++;
+                        }
+                    }
+                }
+
+                pendingRuntimes = runtimeCount;
+                builders = ArrayBuilder<ArrayBuilder<ActiveStatementDebugInfo>>.GetInstance(runtimeCount);
+                builders.Count = runtimeCount;
+
+                // Start execution of the Concord work items.
+                workList.BeginExecution();
+
+                return completion.Task;
+            }
+            finally
+            {
+                // Since we don't own this thread we need to uninitialize the dispatcher. Otherwise if the thread is reused by another bit of code 
+                // trying to call into the Concord API than its InitializeThread will fail.
+                // The work items we queued to the work list above will be run on a dedicated thread maintained by the Concord dispatcher.
+                DkmComponentManager.UninitializeThread(DebuggerComponentIds.ManagedEditAndContinueService);
```

Comment:
Does this reference count internally? Or could this cause other issues?

Reply:
This is the pattern recommended by the debugger team. We need to initialize the thread before we start using any Dkm API and uninitialize it after.

Reply:
The reentrancy is not a problem here in any case since the API isn't called recursively.